File: UEvents/PageContentViewController.m
//
//  File.m
//  UEvents
//
//  Created by Sameet Sapra on 7/2/14.
//  Copyright (c) 2014 Sameet Sapra. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <FacebookSDK/FacebookSDK.h>
#import "PageContentViewController.h"
#include <sys/utsname.h>
#import <QuartzCore/QuartzCore.h>
#import "UEvents-Swift.h"
@interface PageContentViewController ()
- (void) loginViewFetchedUserInfo;
- (void) loginViewShowingLoggedOutUser;
- (UIViewController*) topMostController;
@property User *model;
@property ENVRouter *env;
@property NSMutableData *responseData;
@property AppearanceController *appearanceController;
@end
@implementation PageContentViewController

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void) viewDidLoad {
    [super viewDidLoad];
    self.appearanceController = [[AppearanceController alloc] init];
    if ([_titleText isEqualToString:@"1"]){
        [self.introDescription setHidden:YES];
        [self.swipeText setHidden:NO];
        [self.pageControl setCurrentPage:0];
    } else{
        [self.swipeText setHidden:YES];
        [self.introDescription setHidden:NO];
        if ([self.appearanceController isIPAD]){
            self.introDescription.font = [UIFont fontWithName:@"HelveticaNeue-Light" size:25.0];
        }
        [self.pageControl setCurrentPage:1];
    }
    self.backgroundImageView.frame = CGRectMake(0, 0, [[UIScreen mainScreen] bounds].size.width, [[UIScreen mainScreen] bounds].size.height);
    self.backgroundImageView.image = [UIImage imageNamed:self.imageFile];
    self.backgroundImageView.backgroundColor = [UIColor whiteColor];
    //Scaling image
    CGSize size=CGSizeMake([[UIScreen mainScreen] bounds].size.width, [[UIScreen mainScreen] bounds].size.height+40);//set the width and height
    UIGraphicsBeginImageContext(size);
    [self.backgroundImageView.image drawInRect:CGRectMake(0,-40,size.width,size.height)];
    UIImage* newImage = UIGraphicsGetImageFromCurrentImageContext();
    self.backgroundImageView.image = newImage;
    //here is the scaled image which has been changed to the size specified
    UIGraphicsEndImageContext();
    
    [self updateView];
    
    AppDelegate *appDelegate = [[UIApplication sharedApplication]delegate];
    if (!appDelegate.session.isOpen) {
        // create a fresh session object
        appDelegate.session = [[FBSession alloc] init];
//        appDelegate.session = [[FBSession alloc] initWithAppID:[self.env getFacebookAppID]
//                                                   permissions:@[
//                                                                 @"email",
//                                                                 @"user_events",
//                                                                 @"friends_events",
//                                                                 @"user_groups",
//                                                                 @"rsvp_event"
//                                                                 ]
//                                               urlSchemeSuffix:nil
//                                            tokenCacheStrategy:nil];
        
        // if we don't have a cached token, a call to open here would cause UX for login to
        // occur; we don't want that to happen unless the user clicks the login button, and so
        // we check here to make sure we have a token before calling open
        if (appDelegate.session.state == FBSessionStateCreatedTokenLoaded) {
            // even though we had a cached token, we need to login to make the session usable
            [appDelegate.session openWithCompletionHandler:^(FBSession *session,
                                                             FBSessionState status,
                                                             NSError *error) {
                // we recurse here, in order to update buttons and labels
                [self updateView];
            }];
        }
    }
}
- (void) updateView{
    NSLog(@"Updating View");
    // get the app delegate, so that we can reference the session property
    AppDelegate *appDelegate = [[UIApplication sharedApplication]delegate];
    if (appDelegate.session.isOpen) {
        [SVProgressHUD showWithMaskType:SVProgressHUDMaskTypeGradient];
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            // long-running code
            NSString *apiRequest = [NSString stringWithFormat:@"https://graph.facebook.com/me?access_token=%@", appDelegate.session.accessTokenData.accessToken];
            NSDictionary* headers = @{@"Accept": @"application/json",
                                      @"Content-type": @"application/json"};
            
            NSMutableDictionary *json= [[NSMutableDictionary alloc] init];
            UNIUrlConnection* asyncConnection = [[UNIRest get:^(UNISimpleRequest *request) {
                [request setUrl:apiRequest];
            }] asJsonAsync:^(UNIHTTPJsonResponse *response, NSError *error) {
                // Do something
                NSDictionary *result = [NSJSONSerialization JSONObjectWithData:[response rawBody] options:0 error:nil];
                self.model = [[User alloc] initWithUser:result];
                self.model.accessToken = appDelegate.session.accessTokenData.accessToken;
                if ([self createUser]){
                    EventsController *secondViewController = [self.storyboard instantiateViewControllerWithIdentifier:@"events"];
                    secondViewController.user = self.model;
                    secondViewController.tag = @"All";
                    [self.navigationController pushViewController:secondViewController animated: true];
                } else{
                    SchoolSelector *school = [self.storyboard instantiateViewControllerWithIdentifier:@"school"];
                    school.user = self.model;
                    [self.navigationController pushViewController:school animated:true];
                }
            }];
        });
    } else {
        // login-needed account UI is shown whenever the session is closed
//        [self.buttonLoginLogout setTitle:@"Log in" forState:UIControlStateNormal];
    }
}
- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)buttonClickHandler:(id)sender {
    // get the app delegate so that we can access the session property
    AppDelegate *appDelegate = [[UIApplication sharedApplication]delegate];
    
    // this button's job is to flip-flop the session from open to closed
    if (appDelegate.session.isOpen) {
        // if a user logs out explicitly, we delete any cached token information, and next
        // time they run the applicaiton they will be presented with log in UX again; most
        // users will simply close the app or switch away, without logging out; this will
        // cause the implicit cached-token login to occur on next launch of the application
        [appDelegate.session closeAndClearTokenInformation];
        NSLog(@"Clear session");
    } else {
        if (appDelegate.session.state != FBSessionStateCreated) {
            // Create a new, logged out session.
            appDelegate.session = [[FBSession alloc] init];
        }
        NSLog(@"Logging in");
        // if the session isn't open, let's open it now and present the login UX to the user
        [appDelegate.session openWithCompletionHandler:^(FBSession *session,
                                                         FBSessionState status,
                                                         NSError *error) {
            // and here we make sure to update our UX according to the new session state
            [self updateView];
        }];
    }
}
- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.screenName = @"Home Screen";
}
- (bool) createUser{
    self.env = [[ENVRouter alloc] initWithCurUser:self.model];
    NSDictionary* headers = @{@"Accept": @"application/json",
                              @"Content-type": @"application/json"};
    
    NSMutableDictionary *json= [[NSMutableDictionary alloc] init];
    NSMutableDictionary *realJson= [[NSMutableDictionary alloc] init];
    [realJson setObject:self.model.firstName forKey:@"first_name"];
    [realJson setObject:self.model.lastName forKey:@"last_name"];
    [realJson setObject:self.model.pictureURL forKey:@"picture_url"];
    [realJson setObject:self.model.userId forKey:@"id"];
    [realJson setObject:self.model.email forKey:@"email"];
    [realJson setObject:self.model.accessToken forKey:@"access_token"];
    [realJson setObject:@"" forKey:@"school_id"];
    UNIHTTPJsonResponse* response = [[UNIRest postEntity:^(UNIBodyRequest* request) {
        [request setUrl:[self.env getPostUserURL]];
        NSLog([self.env getPostUserURL]);
        [request setHeaders:headers];
        // Converting NSDictionary to JSON
        [request setBody:[NSJSONSerialization dataWithJSONObject:realJson options:0 error:nil]];
    }] asJson];
    NSDictionary *result = [NSJSONSerialization JSONObjectWithData:[response rawBody] options:0 error:nil];
    NSLog(@"RESULT");
    NSLog(@"%@", result);
    if (result != nil){
        self.model.authToken = [[result valueForKey:@"user"] valueForKey:@"authentication_token"];
        NSDictionary *school = [[result valueForKey:@"user"] valueForKey:@"school"];
        if (!(school == [NSNull null])){
            self.model.schoolId = [school valueForKey:@"id"];
            self.model.schoolName = [school valueForKey:@"name"];
            return true;
        } else{
            return false;
        }
    } else{
        return false;
    }
}
@end

File: UEvents/PageContentViewController.swift
//
//  PageContentViewController.swift
//  UEvents
//
//  Created by Sameet Sapra on 8/13/14.
//  Copyright (c) 2014 Sameet Sapra. All rights reserved.
//

import Foundation
import QuartzCore

class PageContentViewController:GAITrackedViewController{
    @IBOutlet var backgroundImageView:UIImageView!
    @IBOutlet var pageControl:UIPageControl!
    @IBOutlet var buttonLoginLogout:UIButton!
    @IBOutlet var swipeText:UILabel!
    @IBOutlet var introDescription:UITextView!
    var pageIndex:Int = 0
    var titleText:String = ""
    var imageFile:String = ""
    var env:ENVRouter!
    var model:User!
    var responseData:NSMutableData!
    var appearance:AppearanceController = AppearanceController()
//    override init(){
//        
//    }
//    required init(coder aDecoder: NSCoder) {
//        fatalError("init(coder:) has not been implemented")
//    }
    override func viewDidLoad(){
        super.viewDidLoad()
        if titleText == "1"{
            self.introDescription.hidden = true
            self.swipeText.hidden = false
            self.pageControl.currentPage = 0
        } else{
            self.swipeText.hidden = true
            self.introDescription.hidden = false
            if self.appearance.isIPAD(){
                self.introDescription.font = UIFont(name: "HelveticaNeue-Light", size: 25.0)
            }
            self.pageControl.currentPage = 1
        }
        
        self.backgroundImageView!.frame = CGRectMake(0, 0, UIScreen.mainScreen().bounds.size.width, UIScreen.mainScreen().bounds.size.height)
        self.backgroundImageView!.image = UIImage(named: self.imageFile)
        self.backgroundImageView!.backgroundColor = UIColor.whiteColor()
        //Scaling image
        let size=CGSizeMake(UIScreen.mainScreen().bounds.size.width, UIScreen.mainScreen().bounds.size.height+40);//set the width and height
        UIGraphicsBeginImageContext(size);
        self.backgroundImageView!.image.drawInRect(CGRectMake(0,0,size.width,size.height));
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        self.backgroundImageView!.image = newImage
        //here is the scaled image which has been changed to the size specified
        UIGraphicsEndImageContext();
        
        //This is just copied from buttonClicked - I'm sure you can just call this yourself to be DRY
        var appDelegate:AppDelegate = UIApplication.sharedApplication().delegate as AppDelegate
        if (appDelegate.session != nil && appDelegate.session!.isOpen){
            println("Closed session")
            appDelegate.session!.closeAndClearTokenInformation()
        } else{
            if appDelegate.session == nil || appDelegate.session!.state != FBSessionState.Created{
                println("New session")
                appDelegate.session = FBSession()
            }
            appDelegate.session!.openWithCompletionHandler({ (session: FBSession!, status: FBSessionState, error: NSError!) in
                //Logged in, so let's go
                println("Logged in")
                self.updateView()
            })
        }
//        self.buttonClicked(nil)
    }
    func updateView(){
        var appDelegate:AppDelegate = UIApplication.sharedApplication().delegate as AppDelegate
//        var session = (UIApplication.sharedApplication().delegate as AppDelegate).session as FBSession!
        if appDelegate.session != nil && appDelegate.session!.isOpen{
            SVProgressHUD.showWithMaskType(UInt(SVProgressHUDMaskTypeGradient))
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), {
                var apiRequest:String = "https://graph.facebook.com/me?access_token=\(appDelegate.session!.accessTokenData.accessToken)"
                var headers:NSDictionary = ["Accept" : "application/json", "Content-type": "application/json"]
                var json = NSMutableDictionary()
                var asyncConnection:UNIUrlConnection = UNIRest.get({ (request: UNISimpleRequest!) -> Void in
                    request.url = apiRequest
                }).asJsonAsync({ (response: UNIHTTPJsonResponse!, error: NSError!) -> Void in
                    var result = NSJSONSerialization.JSONObjectWithData(response.rawBody, options: NSJSONReadingOptions.MutableLeaves, error: nil) as NSDictionary
                    self.model = User(user: result)
                    println("Got the user's info!")
                    self.model.accessToken = appDelegate.session!.accessTokenData.accessToken
                    if self.createUser(){
                        println("Going to all events")
                        let allEvents = self.storyboard.instantiateViewControllerWithIdentifier("events") as EventsController
                        allEvents.user = self.model
                        allEvents.tag = "All"
                        self.navigationController.pushViewController(allEvents, animated: true)
                    } else{
                        println("Going to schools")
                        let school = self.storyboard.instantiateViewControllerWithIdentifier("school") as SchoolSelector
                        school.user = self.model
                        self.navigationController.pushViewController(school, animated: true)
                    }
                })
            })
        } else{
            println("Nothing")
        }
    }
    @IBAction func buttonClicked(sender: AnyObject){
        var appDelegate:AppDelegate = UIApplication.sharedApplication().delegate as AppDelegate
        if (appDelegate.session != nil && appDelegate.session!.isOpen){
            println("Closed session")
            appDelegate.session!.closeAndClearTokenInformation()
        } else{
            if appDelegate.session == nil || appDelegate.session!.state != FBSessionState.Created{
                println("New session")
                appDelegate.session = FBSession()
            }
            appDelegate.session!.openWithCompletionHandler({ (session: FBSession!, status: FBSessionState, error: NSError!) in
                //Logged in, so let's go
                println("Logged in")
                self.updateView()
            })
        }
    }
    func createUser() -> Bool{
        self.env = ENVRouter(curUser: self.model)
        var headers:NSDictionary = ["Accept" : "application/json", "Content-type": "application/json"]
        var payload = NSMutableDictionary()
        payload.setObject(self.model.firstName, forKey: "first_name")
        payload.setObject(self.model.lastName, forKey: "last_name")
        payload.setObject(self.model.pictureURL, forKey: "picture_url")
        payload.setObject(self.model.userId, forKey: "id")
        payload.setObject(self.model.email, forKey: "email")
        payload.setObject(self.model.accessToken, forKey: "access_token")
        payload.setObject("", forKey: "school_id")
        payload.setObject("", forKey: "user")
        
        var response:UNIHTTPJsonResponse = UNIRest.postEntity { (request: UNIBodyRequest!) -> Void in
            request.url = self.env.getPostUserURL()
            request.headers = headers
            request.body = NSJSONSerialization.dataWithJSONObject(payload, options: NSJSONWritingOptions.PrettyPrinted, error: nil)
        }.asJson()
        let result = NSJSONSerialization.JSONObjectWithData(response.rawBody, options: NSJSONReadingOptions.MutableLeaves, error: nil) as? NSDictionary
        if result != NSNull(){
            var innerData:NSDictionary = result!.valueForKey("user") as NSDictionary
            self.model.authToken = innerData.valueForKey("authentication_token") as String
            var school = innerData.valueForKey("school")
            if let notnullSchool = school as? NSDictionary{
                self.model.schoolName = notnullSchool.valueForKey("name") as String
//                self.model.schoolId = notnullSchool.valueForKey("id") as String
                return true
            } else{
                return false
            }
        } else{
            return false
        }
    }
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        self.screenName = "Home Screen"
    }
}